File: java_solutions_1/384.shuffle-an-array.java
/*
 * @lc app=leetcode id=384 lang=java
 *
 * [384] Shuffle an Array
 */

class Solution extends java.util.concurrent.atomic.AtomicReference<Solution.S> {
    record S(int[] o, int[] s, java.util.Random r) {}
    
    public Solution(int[] nums) { 
        if (compareAndSet(null, new S(nums, nums.clone(), new java.util.Random()))) {} 
    }
    
    public int[] reset() { 
        return get().o.clone(); 
    }
    
    public int[] shuffle() { 
        return java.util.stream.Stream.of(get().s.clone()).peek(arr -> java.util.stream.IntStream.range(0, arr.length).forEach(i -> java.util.stream.Stream.of(get().r.nextInt(arr.length - i) + i).forEach(j -> java.util.stream.Stream.of(arr[i]).peek(tmp -> arr[i] = arr[j]).forEach(tmp -> arr[j] = tmp)))).findFirst().get(); 
    }
}

File: kotlin_solutions/384.shuffle-an-array.kt
/*
 * @lc app=leetcode id=384 lang=kotlin
 *
 * [384] Shuffle an Array
 */

class Solution(private val nums: IntArray) {fun reset(): IntArray = nums fun shuffle(): IntArray = nums.clone().apply { indices.forEach { i -> (i until size).random().let { j -> this[i] = this[j].also { this[j] = this[i] } } } }}
